packages/bff/src/services/authorizationService.ts

```diff
@@ -0,0 +1,149 @@
+import { PagoPAInteropBeClients } from "../providers/clientProvider.js";
+import { config } from "../config/config.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function authorizationServiceBuilder(
+  interopTokenGenerator: InteropTokenGenerator,
+  tenantProcessClient: PagoPAInteropBeClients["tenantProcessClient"],
+  allowList: string[]
+) {
+  const readJwt = async (
+    identityToken: string,
+    logger: Logger
+  ): Promise<{
+    roles: string;
+    sessionClaims: SessionClaims;
+    selfcareId: string;
+  }> => {
+    const verified = await verifyJwtToken(identityToken, logger);
+    if (!verified) {
+      throw tokenVerificationFailed();
+    }
+
+    const decoded = decodeJwtToken(identityToken);
+
+    const {
+      success: validToken,
+      data: token,
+      error,
+    } = AuthToken.safeParse(decoded);
+    if (!validToken) {
+      const claim = error?.errors[0]?.path.join(".");
+      throw missingClaim(claim);
+    }
+
+    if (token.role) {
+      throw genericError("User roles in context are not in valid format");
+    }
+
+    const sessionClaims = SessionClaims.parse(token);
+
+    const selfcareId = token[ORGANIZATION].id;
+
+    return {
+      roles: token[USER_ROLES].join(","),
+      sessionClaims,
+      selfcareId,
+    };
+  };
+
+  const assertTenantAllowed = (selfcareId: string, origin: string): void => {
+    if (
+      !config.tenantAllowedOrigins.includes(origin) &&
+      !allowList.includes(selfcareId)
+    ) {
+      throw tenantLoginNotAllowed(selfcareId);
+    }
+  };
+
+  const buildJwtCustomClaims = (
+    roles: string,
+    tenantId: string,
+    selfcareId: string,
+    tenantOrigin: string,
+    tenantExternalId: string
+  ): CustomClaims => ({
+    [USER_ROLES]: roles,
+    [ORGANIZATION_ID_CLAIM]: tenantId,
+    [SELFCARE_ID_CLAIM]: selfcareId,
+    [ORGANIZATION_EXTERNAL_ID_CLAIM]: {
+      [ORGANIZATION_EXTERNAL_ID_ORIGIN_CLAIM]: tenantOrigin,
+      [ORGANIZATION_EXTERNAL_ID_VALUE_CLAIM]: tenantExternalId,
+    },
+  });
+
+  return {
+    getSessionToken: async (
+      correlationId: string,
+      identityToken: string,
+      logger: Logger
+    ): Promise<string> => {
+      const { sessionClaims, roles, selfcareId } = await readJwt(
+        identityToken,
+        logger
+      );
+
+      const { serialized } =
+        await interopTokenGenerator.generateInternalToken();
+
+      const headers = {
+        "X-Correlation-Id": correlationId,
+        Authorization: `Bearer ${serialized}`,
+      };
+
+      const tenantBySelfcareId =
+        await tenantProcessClient.selfcare.getTenantBySelfcareId({
+          params: { selfcareId },
+          headers,
+        });
+      const tenantId = tenantBySelfcareId.id;
+
+      const tenant = await tenantProcessClient.tenant.getTenant({
+        params: { id: tenantId },
+        headers,
+      });
+
+      assertTenantAllowed(selfcareId, tenant.externalId.origin);
+
```

Comment:
According to the Scala code here is the call where we need to manually call the rate limiter bypassing the middleware because we don't have the tenantId until this point in the code. I will open a task for this, to be done after this PR and the Rate Limiter PR are merged.

Reply:
Here it is: https://pagopa.atlassian.net/browse/IMN-774